Automattic/woocommerce-payments
# Fix setting options on non-production environments

Fixes #6961 

#### Changes proposed in this Pull Request





Custom options support was removed from non-production environments on WooCommerce 7.9.0+. As a result, when making POST requests to the Woo Core Options API via the `updateOptions` function on the frontend of a non-production environment, a 403 response was returned.

This PR fixes the issue by leveraging the [`woocommerce_rest_api_option_permissions` filter](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/src/Admin/API/Options.php#L149) available in the Options API. The PR includes all eight options currently used by frontend `updateOptions` calls in an array that is then merged with the Woo core permissions array via the aforementioned filter.





#### Testing instructions





* Check out this branch.
* All tests should pass.
* Make sure the following two options are deleted from your local client: 
  * `wcpay_frt_discover_banner_settings`
  * `wcpay_fraud_protection_welcome_tour_dismissed`
* Apply the following patch to dump the filtered Woo core option permissions array to `debug.log`: [pr7483.patch](https://github.com/Automattic/woocommerce-payments/files/13394632/pr7483.patch)
* Open your browser inspector and click on the Network tab.
* Navigate to Payments > Overview.
* In the browser inspector network tab look for a request to `options?_locale=user`.
  * The request should have a 200 response.
  * The request payload should be: `{"wcpay_frt_discover_banner_settings": "{\"dontShowAgain\":false}"}`
* Now check your local client's `debug.log` file. You should see the `$permissions` array in the file. Confirm that the following eight options are at the bottom of the array:
  * `wcpay_frt_discover_banner_settings`
  * `wcpay_multi_currency_setup_completed`
  * `woocommerce_dismissed_todo_tasks`
  * `woocommerce_remind_me_later_todo_tasks`
  * `woocommerce_deleted_todo_tasks`
  * `wcpay_fraud_protection_welcome_tour_dismissed`
  * `wcpay_capability_request_dismissed_notices`
  * `wcpay_onboarding_eligibility_modal_dismissed`
* Now, back on the Payments > Overview page, click the Learn More button on the _Enhanced fraud protection for your store_ banner.
* The site should redirect to the Payments > Settings page, then auto-scroll to the bottom of the page and show a welcome tour modal. Click the X in the top right corner to dismiss the modal. 
* Check the browser inspector network tab for a request to `options?_locale=user`.
  * The request should have a 200 response.
  * The request payload should be: `{"wcpay_fraud_protection_welcome_tour_dismissed": true}`
* Now navigate back to Payments > Overview, and click the Dismiss button on the _Enhanced fraud protection for your store_ banner.
* In the browser inspector network tab look for a request to `options?_locale=user`.
  * The request should have a 200 response.
  * The request payload should be: `{"wcpay_frt_discover_banner_settings": "{\"dontShowAgain\":true}"}`
* Lastly, visit the client site dashboard, then return to the Payments > Overview page. The _Enhanced fraud protection for your store_ banner should not render.

No additional unit tests were added to cover this callback method as there are no tests for any of the hook callbacks registered in the `init()` method in `includes/class-wc-payments.php`.

-------------------


**Post merge**
